tests/unit/dispatch/test_lazy_dispatch.py

```diff
@@ -0,0 +1,63 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import pytest
+
+from merlin.dispatch.lazy import lazysingledispatch
+
+try:
+    import tensorflow as tf
+except ImportError:
+    tf = None
+
+
+@lazysingledispatch
+def return_type_name(arg):
+    raise NotImplementedError()
```

Comment:
We tried pretty hard to avoid the need to raise here, but couldn't figure out a way to do it. It sort of falls out as a consequence of wrapping `@singledispatch`. Possible that we could add a utility to define a default method implementation that raises the error in order to enable the syntax you highlighted though.

Reply:
Yeah, there may not be an elegant solution here, but something like this *may* work:

```python
class LazyDispatcher:
    def __init__(self, func_or_name):
        if callable(func_or_name):
            name = func_or_name.__name__
            func = func_or_name
        elif isinstance(func_or_name, str):
            name = func_or_name
            func = None
        else:
            raise TypeError()
        if func is None:
            def _default(*args, **kwargs):
                raise NotImplementedError()
            _default.__name__ = name
            func = _default
        self.dispatcher = singledispatch(func)
        self.func = func
        self._lazy = {}
        self.__name__ = name
 ```

Reply:
I dig it, let's try that